File: welib/vortilib/panelcodes/vortexpoints.py
""" 
Implement a vortex point panel method.

"""
import numpy as np
import pandas as pd
from welib.tools.clean_exceptions import *
from welib.vortilib.panelcodes.panel_tools import airfoil_params, plot_airfoil
# TODO look at welib.airfoils as well.. 


def vp_u(DX, DY, Gamma=1, regParam=0, regMethod=None): 
    """ 
    Induced velocity by one 2D vortex point on one Control Point (CP)
    CP: position of control point
    Pv: position of vortex
    """
    r2 = DX ** 2 + DY ** 2
    tX = -DY
    tY =  DX
    if regMethod is None:
        U = Gamma/(2*np.pi) * tX/r2   # [1] Eq 32.7
        V = Gamma/(2*np.pi) * tY/r2 
    else:                  
        U = Gamma/(2*np.pi) * tX/r2 * (1 - np.exp(- r2 / regParam ** 2))
        V = Gamma/(2*np.pi) * tY/r2 * (1 - np.exp(- r2 / regParam ** 2))
    return U,V


def kutta(M, rhs, iTE=None, verbose=False):
    M2 = M.copy()
    rhs2 = rhs.copy()
    if iTE==0:
        if verbose:
            print('Kutta with iTE=0')
        M2[:,-1] -= M2[:,0]
        M2[:,0:-1] = M2[:,1:]
        M2[:,-1]=-9

        M2[-1,:] -= M2[0,:]
        M2[0:-1,:] = M2[1:,:]
        M2[-1,:]=-9

        rhs2[-1]  -= rhs2[0]  # Subtracting row iTE+1 to col iTE
        rhs2[0:-1] = rhs2[1:] # Shifting indices
        rhs2[-1]   = -9
    elif iTE==-1:
        if verbose:
            print('Kutta with iTE=-1')
        M2[:,0] -= M2[:,-1]
        M2[:,-1]=-9

        M2[-1,:] -= M2[0,:]
        M2[-1,:]=-9

        rhs2[-1]  -= rhs2[0]  # Subtracting row iTE+1 to col iTE
        rhs2[-1]   = -9
    else:
        if verbose:
            print(f'Kutta with iTE={iTE}')
        M2[:,iTE]    -=M2[:,iTE+1]  # Subtracting col iTE+1 to col iTE
        M2[:,iTE+1:-1]=M2[:,iTE+2:] # Shifting indices 
        M2[:,-1]=-9

        M2[iTE,:]     -= M2[iTE+1,:]  # Subtracting row iTE+1 to col iTE
        M2[iTE+1:-1,:] = M2[iTE+2:,:] # Shifting indices
        M2[-1,:]       = -9

        rhs2[iTE]     -= rhs2[iTE+1]  # Subtracting row iTE+1 to col iTE
        rhs2[iTE+1:-1] = rhs2[iTE+2:] # Shifting indices
        rhs2[-1]     = -9

    M2   = M2[:-1,:-1]
    rhs2 = rhs2[:-1]
    return M2, rhs2

def kutta_revert(sol, iTE=None, start='LE'):
    """" revert the Kutta condition """
    sol2 = np.zeros(len(sol)+1)
    if iTE==0:
        sol2[-1] = -sol[0]
        sol2[:-1]= sol
    elif iTE==-1:
        sol2[-1]= -sol[0]
        sol2[:-1] = sol
    else:
        sol2[:iTE+1] =  sol[:iTE+1]
        sol2[iTE+1]  = -sol[iTE]
        sol2[iTE+2:] = sol[iTE+1:]
    return sol2

def backDiagonalCorrection(M, ds):
    """ Apply backdiagonal correction to enforce zero internal circulation"""
    m = M.shape[0]
    for i in range(m):
        vsum=0
        for j in range(m):
            if j!=m-i-1:
                vsum+= M[j,i]*ds[j]
        M[m-i-1, i] = -vsum/ ds[m-i-1]
    return M


def panel_solve_vps(XP, YP, Ux, Uy, lift=True, iTE=0, curv_method='Menger', backDiagCorr=True, verbose=False):
    """ 
    """

    # --- Geometry
    normals, tangents, mids, ds, curvature, ax =  airfoil_params(XP, YP, plot=False, ntScale=0.3, curv_method=curv_method)

    if verbose:
        print('ds  ', ds)
        print('crv ', curvature)
        #print('cds2', curvature*ds*2)

    # --- Right hand side
    # We implement the "Dirichlet" condition, no flow tangential
    rhs = -(Ux*tangents[:,0] + Uy*tangents[:,1])

    # --- Build matrix
    m = len(XP)-1
    M = np.zeros((m,m))
    for i in range(m):
        for j in range(m): # 
            if i==j:
                # Self weight and curvature
                # See Lewis
                M[i,i] = -0.5 - curvature[i]*ds[i]/(4*np.pi)
            else:
               ri=mids[i,:] # 
               rj=mids[j,:] # 
               dr = ri-rj
               u,v = vp_u(dr[0], dr[1], Gamma=1, regParam=0, regMethod=None)
               M[j,i] = (u*tangents[j,0] + v*tangents[j,1]) * ds[i]

    # --- Back diagonal correction
    # See Lewis
    if lift:
        if backDiagCorr:
            M = backDiagonalCorrection(M, ds)

    # --- KUTTA condition
    if lift: 
        M, rhs = kutta(M, rhs, iTE)

    # --- SOLVE
    gammas_r = np.linalg.solve(M, rhs)

    if lift: 
        gammas = kutta_revert(gammas_r, iTE=iTE)
        rhs    = kutta_revert(rhs, iTE=iTE)
    else:
        gammas = gammas_r


    # --- Outputs
    out = {}
    # Geometry
    out['x']    = XP
    out['y']    = YP
    out['theta'] = np.arctan2(YP, XP)
    out['ds']   = ds
    out['n']    = normals
    out['t']    = tangents
    out['CP']   = mids
    out['VP']   = mids
    out['curv'] = curvature
    out['theta_CP'] = np.arctan2(mids[:,1], mids[:,0])
    out['rhs']  = rhs

    # --- Output: Velocity at wall
    # TODO
    #Vtheta = gammas
    Vwall = np.zeros_like(mids)
    # NOTE: using the fact that tangent condition should be satisfied
    Vwall[:,0]=gammas*tangents[:,0] + 0*Ux
    Vwall[:,1]=gammas*tangents[:,1] + 0*Uy
    out['Vwall'] = Vwall

    # --- Output: Cp
    out['Cp'] = 1-(Vwall[:,0]**2+Vwall[:,1]**2)/(Ux**2+Uy**2)


    # --- Output: Lift coefficient
    # TODO
    # LEwis Eq. 2.30

    # --- Output: Stream function
    #psi = Ux*y - Uy*x + 1/(2*np.pi) * sum(gamma_i ds_i ln(r_ii))

    return gammas, out



if __name__ == '__main__':
    case = 'file'
#     case = 'cylinder'
#     case = 'ellipse_lift'
    alpha=0
    ge=None
    curv_method='zero'
#     curv_method='Menger'
    backDiagCorr=False
#     backDiagCorr=True
    coords_filename =None
    if case=='file':
        U0    = 1
        alpha = 9*np.pi/180
#         coords_filename = 'FFA-W3-301-coords.csv'
        coords_filename = 'geom-KarmanTrefftz_less.csv'
#         coords_filename = 'geom-KarmanTrefftz.csv'
#         coords_filename = 'Diamond-coords.csv'
        coords_filename = 'NACA0012-n399-coords.csv'
        # --- Read data
        df = pd.read_csv(coords_filename)
        XP = df['x'].values
        YP = df['y'].values
        lift=True
    elif case=='cylinder':
        m=21
        U0    = 1
        R=2
        theta   =-np.linspace(0,2*np.pi, m+1)
        XP = R*np.cos(theta)
        YP = R*np.sin(theta)
        dtheta=theta[1]-theta[0]
        theta_mid   = theta[:-1]+dtheta/2
        ge =  2.0*U0*np.sin(theta_mid) # Gamma_exact = Utheta at CP
        lift=False
    elif case=='ellipse_lift':
        m=300
        U0    = 1
        major=1
        ratio=0.05
        lift=True
        alpha = 30*np.pi/180
        theta   =-np.linspace(0,2*np.pi, m+1)
        abyr= np.sqrt((1.-ratio)/(1.+ratio))# ! see Lewis p 50
        XP = 0.5*major       *(np.cos(theta))
        YP = 0.5*major*ratio*  np.sin(theta)
        dtheta=theta[1]-theta[0]
        theta_mid   = theta[:-1]+dtheta/2
        f = np.sqrt(1.0+ (abyr**4) - 2.*(abyr**2)*np.cos(2.*theta_mid) )
        ge =  (2.0*U0*np.sin(theta_mid-alpha)+2*np.sin(alpha))/f          # See Lewis p50


    Ux    = U0*np.cos(alpha)
    Uy    = U0*np.sin(alpha)

    gammas, out = panel_solve_vps(XP, YP, Ux, Uy, lift=lift, iTE=0, curv_method=curv_method, verbose=True, backDiagCorr=backDiagCorr)
    print(gammas)
    VP = out['VP']

    #plot_airfoil(XP, YP, Uwall=out['Vwall'], ntScale=0.1, UScale=0.8)
    plot_airfoil(XP, YP, nt=True, ntScale=0.1, UScale=0.8)
    
#     print('MaxError ',np.max(np.abs(ge-gn)))
#     print('')
#     print(f'Case: {iCase} - Curv Method: {curv_method} - iCoord: {iCoord} - backCorr:{backCorr} - m:{m}')
    import matplotlib.pyplot as plt
    fig,axes = plt.subplots(1, 2, sharey=False, figsize=(6.4,4.8)) # (6.4,4.8)
    fig.subplots_adjust(left=0.12, right=0.95, top=0.95, bottom=0.11, hspace=0.20, wspace=0.20)
    ax=axes[0]
    if ge is not None:
        ax.plot(VP[:,0], ge    , 'k-', label='Theory')
    ax.plot(out['VP'][:,0], gammas, '--', label='gammas')
    ax=axes[1]
    if ge is not None:
        ax.plot(out['VP'][:,0], 1-ge**2/U0**2, '--', label='Theory')
    ax.plot(out['VP'][:,0], out['Cp'], '--', label='Cp')
    ax.legend()

    if coords_filename is not None:
        if 'Karman' in coords_filename:
            M = np.column_stack((out['VP'][:,0], out['Cp']))
            df = pd.DataFrame(data=M,columns=['x','Cp'])
            df.to_csv('CP-KarmanTrefftz_5_num_py.csv',sep=',', index=False)


#     ax.plot(theta_mid, gammas, '--', label='gammas')
#     ax.plot(theta_mid, ge    , '-', label='gammas')
#     ax.set_xlabel('phi')
#     ax=axes[1]
#     ax.plot(theta_mid*180/np.pi,rhs, label='rhs')
#     ax.plot(theta_mid*180/np.pi,ge , label='ge')
#     ax.plot(theta_mid*180/np.pi,gn ,'--', label='gn')
#     ax.set_xlabel('theta')
#     ax.set_ylabel('')
#     ax.legend()
#     # plt.show()


    plt.show()

File: welib/vortilib/panelcodes/matlab/vortexpoints.m
%% Initialization
clear all; close all; clc; %restoredefaultpath; %addpath(genpath())

%% Functions

function [U, V] = vortex_flow(DX, DY, Gamma, regParam, regMethod)
    % Induced velocity by one 2D vortex point on one Control Point (CP)
    % INPUTS:
    %   - DX, DY: vector from vortex to the control point (array-like)
    %   - Gamma: vortex strength
    %   - regParam: regularization parameter (default value: 0)
    %   - regMethod: regularization method (default value: [])
    % OUTPUTS:
    %   - U, V : Cartesian velocities (size of DX)
    if nargin < 4
        regParam = 0;
        regMethod = 0;
    end
    r2 = DX.^2 + DY.^2;
    tX = -DY;
    tY =  DX;
    bOK = r2 > 1e-8; % Avoid singularity at r=0
    U = zeros(size(DX));
    V = zeros(size(DY));
    if regMethod==0
        U(bOK) = Gamma/(2*pi) * tX(bOK)./r2(bOK); 
        V(bOK) = Gamma/(2*pi) * tY(bOK)./r2(bOK); 
    else
        U(bOK) = Gamma/(2*pi) * tX(bOK)./r2(bOK) .* (1 - exp(- r2(bOK) / regParam^2));
        V(bOK) = Gamma/(2*pi) * tY(bOK)./r2(bOK) .* (1 - exp(- r2(bOK) / regParam^2));
    end
end

function out = VP_panel_solve(x, y, Vinf_x, Vinf_y, hasLift)
    % Solve the flow about an airfoil using the vortex point method.
    % INPUTS:
    %  -x , y : airfoil points, assumed to go from TE to LE clockwise [m]
    %  -Vinf_xy: components of the freestream velocity in the x and y direction [m/s]
    % OUTPUTS:
    %  - out: storage for multiple variables, like Cp

    % Compute geometry (mid points, normals, tangents)
    P     = [x(:), y(:)]                   ;
    mids  = (P(1:end-1,:) + P(2:end,:)) / 2;
    dP    = P(2:end,:) - P(1:end-1,:)      ;
    ds    = sqrt(sum(dP.^2, 2))            ;
    t_hat = dP ./ ds                       ;
    n_hat = [-t_hat(:,2), t_hat(:,1)]      ;

    % Right hand side
    % We will implement the "Dirichlet" condition, no flow tangential "inside" of the body
    rhs = 0*x; % TODO Implement the right hand side: rhs = -Vinf . t
    rhs = -(Vinf_x*t_hat(:,1) + Vinf_y*t_hat(:,2));

    % Build matrix
    nCP = length(x)-1;  % Number of Control points
    nV = length(x)-1;   % Number of vortices
    CP = mids;          % Position of control points
    VP = mids;          % Position of vortices
    M = zeros(nCP,nV);
    for i = 1:nCP
        for j = 1:nV
            if i == j
                M(i,i) = 0.5/ds(j); % Exactly on panel, panel strength is half the velocity jump
            else
                ri = CP(i,:); % control point position
                rj = VP(j,:); % vortex point position
                % Induced velocity Vij from one unit vortex point at one control point.
                u = 0; % TODO Implement it
                v = 0; % TODO Implement it
                % Velocity projeted against the normal vector M[i,j] = Vij . t
                M(i,j) = 0; % TODO

                dr = ri - rj;
                [u, v] = vortex_flow(dr(1), dr(2), 1);
                % Velocity projected against the normal vector M(i,j) = Vij . t
                M(i,j) = (u*t_hat(i,1) + v*t_hat(i,2));
            end
            % Scale by length at the end (we formulate in terms of gammas)
            M(i,j) = M(i,j) * ds(j);
        end
    end
    if hasLift
        % Bonus: insert Kutta boundary condition here
    end

    % Solve (invert the system)
    gammas = M\rhs;
    if hasLift
        % Bonus: revert Kutta boundary condition here
    end
    Gammas = gammas.*ds;

    % Outputs
    out.VP = VP;            % Vortex points
    out.M = M;              % System matrix
    out.rhs = rhs;          % Right hand side
    out.Gammas = Gammas;    % Vortex points intensities

    % Velocity at wall (uses the fact that the vortex sheet creates a velocity jump)
    Vwall = zeros(size(mids));
    Vwall(:,1) = -(Gammas./ds).*t_hat(:,1);
    Vwall(:,2) = -(Gammas./ds).*t_hat(:,2);
    out.Vwall = Vwall;

    % Cp
    out.Cp = 0; % TODO Implement Cp using Vwall
    out.Cp = 1 - (Vwall(:,1).^2 + Vwall(:,2).^2) / (Vinf_x^2 + Vinf_y^2);
end



%% Main parameters
case_option = 'file';     % Choose between a file or the cylinder case
% case_option = 'cylinder'; % Choose between a file or the cylinder case
U0    = 1;                % Free stream [m/s]
alpha = 0*pi/180;         % Angle of attack [rad]

% Setup case
if strcmp(case_option, 'file')
    coords_filename = '../data/Diamond-coords.csv';
    data = xlsread(coords_filename);
    xa = data(:, 1); % Coordinates of airfoil
    ya = data(:, 2);
    Cp_theory = ones(length(xa)-1,1)*0.452
    hasLift = abs(alpha)>0;
elseif strcmp(case_option, 'cylinder')
    m         = 100;                % Number of panels
    R         = 2;                 % Cylinder radius [m]
    Gamma     = -4*pi*U0*R*0;      % Circulation - NOTE: Not available yet
    theta     = -linspace(0,2*pi, m+1) + asin(Gamma/(4*pi*U0*R));
    xa        = R*cos(theta);
    ya        = R*sin(theta);
    Cp_theory = 1-4*(sin((theta(1:end-1)+theta(2:end))/2) - Gamma/(4*pi*U0*R)).^2;
    hasLift = abs(Gamma)>0;
end

%% Derived parameters
Vinf_x = U0*cos(alpha);
Vinf_y = U0*sin(alpha);

%% Use the vortex panel method to find the solution
out = VP_panel_solve(xa, ya, Vinf_x, Vinf_y, hasLift);


if strcmp(case_option, 'file')
    fprintf('M matrix:\n')
    disp(out.M)
end


%% Plot results
figure; hold all;
plot(out.VP(:,1), Cp_theory, 'k-', 'DisplayName', 'Theory');
plot(out.VP(:,1), out.Cp, ':o', 'DisplayName', 'Numerical');
ylim([-4., 1.1]);
xlabel('x [m]');
ylabel('Cp [-]');
legend();
